### ISSUE TITLE
Reduce the scanning area?

### ISSUE THREAD

**Issue by blackthornecl** [jpxue/Overwatch-Aim-Assist#10]
How to reduce the area scanning to center of screen and not entire screen?

**Comment by s84088722** [comment]
I think it is good idea!

**Comment by blackthornecl** [comment]
Any idea?

**Comment by jpxue** [comment]
Can't really answer that off the top of my mind.
In capture.cpp: I'm assuming that you'll need to change some of the arguments in CreateCompatibleBitmap, BitBlt, GetDIBits. sWidth and sHeight also need to be changed to your FOV settings.
The BITMAPINFO struct will also have to be assigned manually (you'll have to manually assign size, width, height, planes, bitcount, compression, sizeimage - used to do this manually up until [this](https://github.com/jpxue/Overwatch-Aim-Assist/commit/6b9fdced55d6738de9e2c56037b2ce8891209847) commit). 
Currently I use GetDIBits to assign the BITMAPINFO struct (by pointing to it and by passing a null as lpvBits) - you'll have to omit this before using GetDIBits again to obtain the RGBQUADs (the actual pixel data).

I cannot tell you for everything for certain though unless I try it myself and report back.

**Comment by islanq** [comment]
Forgive me if i'm being ignorant but instead of manually adding all Bitmap args what is wrong with just changing scan location in findPlayer()?

something like:
`int scanOffset = 250;
for (int i = (width/2)-scanOffset; i < (width/2)+scanOffset; i++)`

Either that or instead of starting from the top left why not start from the center outward to left and or right. This way if it already has a target it wont be trying to switch to one farther left. 

Also, the findPlayer function as whole has WAY too much going on in it. The find health bar and the actual pixel scanning iteration(s) work should be refactored so we can experiment with different techniques. I'm not even entirely sure what "length" technically is.

**Comment by jpxue** [comment]
The point of capturing a smaller screenshot would make more sense since screen capturing is what takes up most time (40ms) and not findPlayers (8ms); thus it would be the more efficient optimization.

Yeah, I was wondering how much starting from the centre helps but it would over-complicate the main algorithm + for something which takes 7-8ms the result wouldn't be noticeable. The benefit of starting from top left and moving to the right is that we practically always find the origin of the health bar (i.e. the top left corner of the health bar itself); if not the algorithm which calculates the health bar would need to be changed.

I agree that the findPlayer function should be split into several discrete functions because it is currently the only function with one big chunk of code and is a little bit of a pain to read.
It is also a pain in the ass because bitblt returns a 1D array (screenshots are 2D in nature) and to make things worse, the screenshot is inverted upside-down (this can be addressed by reversing one of the BITMAPINFOHEADER fields which I did not know at the time of writing).

length is the number of RGBQUADs starting from the address pointed to by pixels. I put in a comment in screenshot.h interface file.

**Comment by islanq** [comment]
You bring up some good points. I didn't run any runtime benchmarks on the code but on a quad core i7 it's pretty snappy (for me anyway). 

Currently i don't mind capping the whole screen but i'd like to be able to easily identify where the X coord and Y coord start. For example. sometimes i'll be locked onto a target shooting and if i happen to damage another player next to that person exposing their health bar (player to the left or up and to the left) my aim shifts over to start focusing the unintended target. 

Perhaps adding a function that only engages the health bar closest to the center of the screen until it is gone. Or possibly being able to reduce the area in which it is looking for the target on the X axis to achieve a similar result.

PS. When the game first came out i wrote a quick scanner with OpenCV that i seem to remember benching in around 20ms or so. I'll check it out again to see if it's more efficient or send it your way for you to look at if you'd like. Grabbing the screenshot straight to OpenCV mat container -> HSV color space -> dilate / erode produces binary nice blobs for only the characters.

**Comment by jpxue** [comment]
I tried OpenCV using pretty much your approach i.e. screen capture using GDI followed by HSV conversion -> inRange -> findContours (to find the borders and calculate the coordinates). Excluding screen capturing I think it took ~30ms or so which is also good (I'm on an old i7 - 1st gen).

Yeah, it's alternating between two targets because it is capturing the borders of both targets while scanning under the health bar (because I made it scan a little bit more wider than the health bar itself for large targets such as winston). I didn't see it coming because I programmed this in the training range (a perfect scenario) but I think it can be solved by just scanning for just a small width range under the middle of the health bar itself.
